include/nn/util/util_BitFlagSet.h

```diff
@@ -5,74 +5,235 @@
+    }
+
+    template <typename FlagT>
+    BitFlagSet& Set(bool isOn) const {
+        return SetImpl(FlagT::StorageIndex, FlagT::StorageMask, isOn);
+    }
+
+    int GetCount() const { return N; }
+
+    bool Test(int index) const { return TestImpl(GetStorageIndex(index), MakeStorageMask(index)); }
+
+    template <typename FlagT>
+    bool Test() const {
+        return TestImpl(FlagT::StorageIndex, FlagT::StorageMask);
+    }
+
     template <int BitIndex>
     struct Flag {
-    public:
         static const int Index = BitIndex;
+        // todo: figure out how to make this not break past 63
         static const BitFlagSet Mask = 1 << Index;
 
     private:
         static const int StorageIndex = BitIndex / StorageBitCount;
-        static const StorageT StorageMask = 1 << Index % StorageBitCount;
+        static const StorageT StorageMask = StorageT(1) << BitIndex % StorageBitCount;
     };
 
-    BitFlagSet operator~() const;
-    BitFlagSet operator&(const BitFlagSet&) const;
-    BitFlagSet operator^(const BitFlagSet&) const;
-    BitFlagSet operator|(const BitFlagSet&) const;
-    BitFlagSet& operator&=(const BitFlagSet&);
-    BitFlagSet& operator^=(const BitFlagSet&);
-    BitFlagSet& operator|=(const BitFlagSet&);
-    bool operator==(const BitFlagSet&) const;
-    bool operator!=(const BitFlagSet&) const;
-    bool operator[](int) const;
-    Reference operator[](int);
-    bool IsAnyOn() const;
-    int CountPopulation() const;
-    BitFlagSet& Flip(int);
-    BitFlagSet& Flip();
-    bool IsAllOn() const;
-    bool IsAllOff() const;
-    BitFlagSet& Reset();
-    BitFlagSet& Reset(int);
-    BitFlagSet& Set();
-    BitFlagSet& Set(int, bool);
-    int GetCount() const;
-    bool Test(int) const;
-
 private:
-    BitFlagSet& SetImpl(int, StorageT, bool);
-    bool TestImpl(int, StorageT) const;
-    void Truncate();
-    void TruncateIf(std::true_type);
-    void TruncateIf(std::false_type);
-    static int GetStorageIndex(int);
-    static StorageT MakeStorageMask(int);
+    BitFlagSet& SetImpl(int storageIndex, StorageT storageMask, bool isOn) {
+        if (isOn) {
+            _storage[storageIndex] |= storageMask;
+        } else {
+            _storage[storageIndex] &= ~storageMask;
+        }
+        return *this;
+    }
+
+    bool TestImpl(int storageIndex, StorageT storageMask) const {
+        return _storage[storageIndex] & storageMask;
+    }
+
+    // this is probably used to to reset the unused bitflags in the last StorageT
+    void Truncate() { TruncateIf(std::integral_constant<bool, N % StorageBitCount>()); }
+
+    // todo: figure out an implementation to keep only the valid flags in the last StorageT
+    void TruncateIf(std::true_type) {}
+    void TruncateIf(std::false_type) {}
+
+    static int GetStorageIndex(int index) { return index / StorageBitCount; }
+    static StorageT MakeStorageMask(int index) { return StorageT(1) << index % StorageBitCount; }
```

Comment:
Could you add some parens here to clarify the intended operator precedence order?